keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -91,242 +92,324 @@ with the unreasonable development overhead of a webhook.
-arrays and objects. It recursively applies to the whole subtree:
+*Example 1 - Immutable*:
+
+```yaml
+properties:
+  foo:
+    type: string
+    x-kubernetes-mutability: Immutable | AddOnly | RemoveOnly
+```
+
+Allowed for AddOnly, disallowed otherwise:
+- `{}` → `{foo:"a"}`
+
+Allowed for RemoveOnly, disallowed otherwise:
+- `{foo:"a"}` → `{}`
+
+Disallowed:
+- `{foo:"a"}` →	`{foo:"b"}`
+
+#### Mutable array, immutable items, list type list
+
+The items with their respective index in the array are immutable. Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+
+*Example 2*:
 
 ```yaml
 properties:
   foo:
-    type: array
     items:
-      x-kubernetes-immutable: true
-      type: string
+      type: object
+      properties: ...
+      x-kubernetes-mutability: Immutable
 ```
 
-means that the `foo` array itself is mutable, but the items are not. This means that
-items can be deleted and added, but at each existing index the string values are
-immutable.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[a]}` → `{foo:[a,b]}` (value of key 0 is unchanged)
+- `{foo:[a]}` →	`{foo:[]}`
+
+Disallowed:
+- `{foo:[a]}` → `{foo:[b]}`
+- `{foo:[a]}` → `{foo:[b,a]}` (value of key 0 is changed)
+
+#### Mutable array, immutable items, list type map
+
+The key-value pairs in the array are immutable, set set of keys is not. Hence, addition and removal of key-values pairs
+
+*Example 3*:
 
 ```yaml
 properties:
   foo:
-    type: object
-    x-kubernetes-immutable: true
+    x-kubernetes-list-type: map
+    x-kubernetes-list-map-keys: ["k"]
+    items:
+      type: object
+      properties: ...
+      x-kubernetes-mutability: Immutable
 ```
 
-means that the whole `.foo` object is immutable. It cannot be removed or set.
+Allowed:
+- `{}` → `{foo:...}`
+- `{foo:...}` → `{}`
+- `{foo:[{k:a}]}` → `{foo:[{k:a},{k:b}]}` (value of key a is unchanged)
+- `{foo:[{k:a}]}` → `{foo:[]}`
+- `{foo:[{k:a}]}` → `{foo:[{k:b}]}`
+- `{foo:[{k:a},{k:b}]}`	→ `{foo:[{k:b},{k:a}]}`	(values of key a and b unchanged)
+
+Disallowed:
+- `{foo:[{k:a,v:1}]}` → `{foo:[{k:a,v:2}]}` (value of key a is changed)
+
+**TODO:** removal+addition = change. The former is allowed, the latter disallowed. 
+
+#### Mutable array, immutable items, list type set
 
-If one wants to make only field immutable, but allow to delete an existing object
-or set it if unset, the following is used:
+Sets are maps with the whole value as key. Hence, addition and removed of any value in the set is allowed.
+
+*Example 4*:
 
 ```yaml
 properties:
   foo:
-    type: object
-    properties:
-      "x":
-        type: string
-        x-kubernetes-immutable: true
-      "y":
-        type: string
-        x-kubernetes-immutable: true
+    x-kubernetes-list-type: set
+    items:
+      type: string
+      x-kubernetes-mutability: Immutable
 ```
 
-for all fields `x` and `y` of `foo`.
+Allowed:
```

Comment:
Since `x-kubernetes-list-type: set` value must be a scalar (or another atomic type), what would be not allowed here? can we have an example with
```
    x-kubernetes-list-type: set
    items:
        type: object
        x-kubernetes-mutability: Immutable
```

Reply:
```
    x-kubernetes-list-type: set
    items:
        type: object
        x-kubernetes-mutability: Immutable
        x-kubernetes-map-type: atomic
        additionalProperties:
          type: string
```

Reply:
added